File: Medium/1428-jump-game-iii/jump-game-iii.py
class Solution:
    def solve(self, idx, arr, visited, n):
        if idx<0 or idx>=n:
            return False

        if arr[idx] == 0:
            return True

        if visited[idx] is True:
            return False

        visited[idx]=True
        left=self.solve(idx-arr[idx], arr, visited, n)
        right=self.solve(idx+arr[idx], arr, visited, n)
        return left or right
    
    def canReach(self, arr: List[int], start: int) -> bool:
        n=len(arr)

        visited=[False]*n
        return self.solve(start, arr, visited, n)

File: Medium/1428-jump-game-iii/jump-game-iii.java
class Solution{
    boolean solve(int idx, int[] arr, boolean[] visited, int n){
        if(idx<0 || idx>=n)
            return false;

        if(arr[idx] == 0)
            return true;

        if(visited[idx])
            return false;

        visited[idx]=true;
        boolean left=solve(idx-arr[idx], arr, visited, n);
        boolean right=solve(idx+arr[idx], arr, visited, n);
        return left || right;
    }

    public boolean canReach(int[] arr, int start){
        int n=arr.length;

        boolean[] visited=new boolean[n];
        return solve(start, arr, visited, n);

    }
}